### DISCUSSION TITLE
Expose prometheus metrics for log messages by level

### DISCUSSION THREAD

**Discussion by weeco** [grafana/grafana#37035] [General]
**What would you like to be added**:

Add metric series like `log_messages_total{level="info"}` which show the number of log messages printed on that specific log level.

**Why is this needed**:

Having the number of log messages by log level available as metric allows me as Grafana operator to configure alerts if the number of error messages suddenly increase. This could be the case some alert notifications can not be sent anymore.

**Comment by torkelo** [comment]
Not sure this is a Grafana request, but a loki ingest feature request?

**Comment by weeco** [comment]
@torkelo No. I am asking for Grafana to introduce a new metric called `log_messages_total`. Whenever Grafana prints log messages it should increment this counter by one for the given log level so that the Grafana container exposes the following metrics:

```
log_messages_total{level="debug"} 5400
log_messages_total{level="info"} 981
log_messages_total{level="warn"} 0
log_messages_total{level="error"} 0
```

Now I can add a panel which monitors the number of error messages logged in Grafana. If there's a sudden spike I can be alerted on this metric so that someone checks the log messages to see what's going on. Is it more clear now?

**Comment by torkelo** [comment]
Ah, sorry for the misunderstanding! A good feature request!

**Comment by conorevans** [comment]
Hi @torkelo this seems pretty reasonable to me - can you assign it please? 🙂

**Comment by conorevans** [comment]
So I looked at it - the code itself was predictably trivial, but I forgot that there was likely gonna be a circular import - `pkg/infra/logs` now has an import for `pkg/infra/metrics` which has several files that import `/pkg/setting` which itself has many files that import `pkg/infra/logs`. While one file referencing `/pkg/setting` was easy to update, the other two seem less trivial.

Overall, I'm not 100% which route to take - I haven't done much Go refactoring.

**Comment by pkolyvas** [comment]
I'm going to move this to GitHub discussions while we work on a path forward.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
